### PR TITLE
feat(android)(game): integrate finished-game result screen and winner flow

### PR DESCRIPTION
**Pull request by vanidrei**
# PR Summary: Frontend End Game Screen Integration

Feature Status
Scope Platform Domain
Affects Affects Affects

- Closes #576 
- Closes #577 
- Closes #578 
- Closes #579 
- Closes #580 
- Closes #581
- Closes #582 
- Closes #583 
- Closes #818 
- Closes #817 
- Closes #816 
- Closes #815 
- Closes #814 
- Closes #813 

## Summary

Implements the complete game-end flow: detecting when a game finishes, building a result model, navigating to a dedicated result screen, and showing a full leaderboard with player stats and placement badges.

---

### Effect & Routing
- Added `GameEffect.NavigateToResult(winnerUserId: String)` to the sealed effect class
- Added `Routes.GAME_RESULT = "game_result"` route constant
- Added `GameTestTags` constants for all result screen elements (`RESULT_SCREEN`, `RESULT_TITLE`, `RESULT_WINNER_NAME`, `RESULT_LEADERBOARD`, `RESULT_HOME_BUTTON`)

### State
- Added `GameResultPlayerSummary` model with full stat fields: `tilesPlayed`, `meldsCreated`, `turnsCompleted`, `pointsFromTiles`, `remainingTiles`, `penaltyPoints`, `finishPosition`, `isStillPlaying`
- Added `GameResultUiModel` with `winnerUserId`, `players`, `matchDuration`, `totalTurns`, `finishedTimestamp`
- Added `gameResult: GameResultUiModel?` and `elapsedSeconds: Int` to `GameUiState`
- Removed `winnerUserId: String?` from `GameUiState` (superseded by `GameResultUiModel`)

### ViewModel
- Added `handleFinishedGame(winnerUserId)` — builds `GameResultUiModel`, captures match duration, emits `NavigateToResult` effect
- Added `hasEmittedGameResultNavigation` deduplication guard — prevents double navigation when multiple signals arrive simultaneously
- Game-end detected from **three sources** for resilience:
  - `GameEvent.Ended` — primary, winner ID from backend
  - `GameEvent.Updated` with `status == FINISHED` — fallback
  - `applyGameState()` with `status == FINISHED` — handles reconnect/resume
- Added live match timer: starts on `loadGame()`, increments `elapsedSeconds` every second, captured as `matchDuration` on game end

### GameResultScreen (new file)
- Dark purple gradient themed screen
- **Header**: back button, game ID, play time
- **Hero section**: GAME OVER badge, trophy icon, YOU WIN! / YOU FINISHED title, players-finished pill
- **Stats row**: Duration, Moves, Points, Melds (current player)
- **Leaderboard**: rank badge, avatar placeholder, name, placement badge, stats line (tiles left · melds · penalty · bonus), score
- **Placement badges**: 🏆 1st Place (gold) · 🥈 2nd Place (silver) · 🥉 3rd Place (bronze)
- **Bottom stats row**: Tiles Played, Sets Created, Avg Turn
- **Bottom actions**: Next Round, Share (wired later), Home

### Navigation
- `HomeGraph` collects `NavigateToResult` effect → navigates to `game_result/{winnerId}`
- `GameResultScreen` reuses `GameViewModel` from the game back-stack entry so full player data is available without re-fetching
- **Next Round** navigates back to `WaitingRoom` with the same `lobbyId` (same players, new round)
- **Home** navigates to `HomeScreen` clearing the back stack

### Other
- Increased OkHttp connect/read/write timeouts from 30s → 60s (Render.com free-tier cold start)
- `GameScreen` header timer replaced hardcoded `"3:45"` with live `elapsedSeconds` formatted as `M:SS`
- Added temporary 🏁 debug button on `GameScreen` to trigger `handleFinishedGame()` without completing a full game *(marked DEBUG ONLY — remove before release)*

---

## Notes

- Backend stat fields (`tilesPlayed`, `meldsCreated`, `turnsCompleted`, `penaltyPoints`) are currently placeholder `0` — will be mapped once the backend provides them per-player
- Winner is determined by `GameEvent.Ended.winnerUserId` from the backend; fallback ranking uses highest score
- The 🏁 debug button in `GameScreen` must be removed before merging to `main`

---

## Test plan

- [ ] Start a game with two players, tap 🏁 on one device → result screen appears on both
- [ ] Confirm YOU WIN! shown on winner's device, YOU FINISHED on loser's
- [ ] Confirm placement badges (🏆 1st, 🥈 2nd) are correct
- [ ] Tap **Next Round** → both players land in WaitingRoom for the same lobby
- [ ] Tap **Home** → navigates to HomeScreen, back stack is cleared
- [ ] Kill and reopen the app mid-game → game-end detection via `applyGameState()` still navigates to result screen
- [ ] Verify match timer counts up correctly and is frozen on result screen

### CODE REVIEW

File: apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt [hunk 1]
```diff
@@ -146,13 +149,52 @@ fun NavGraphBuilder.homeGraph(
...
+                            navController.navigate("${Routes.GAME_RESULT}/${effect.winnerUserId}")
                     }
                 }
             }
 
             GameScreen(viewModel = vm)
         }
 
+        composable("${Routes.GAME_RESULT}/{winnerId}") { backStackEntry ->
+            val winnerId = backStackEntry.arguments?.getString("winnerId").orEmpty()
```
Comment (kingjulien1) [inline apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt]: I think this should be tightened one way or the other.

Right now we navigate with `winnerId`, but the result screen does not actually use that route argument and instead renders from `uiState.gameResult`. Since Sonar is already flagging the local variable as unused, I’d either remove the route parameter entirely or make it part of the actual screen contract.

(Personally, I would remove the route parameter entirely 😉)

File: apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+private val ResRed           = Color(0xFFFF5252)
+private val ResGray          = Color(0xFF9E9E9E)
+private val ResStillPlaying  = Color(0xFF4ECDC4)
+
+@Composable
+fun GameResultScreen(
+    gameResult: GameResultUiModel?,
+    currentUserId: String? = null,
+    onNavigateHome: () -> Unit = {},
+    onNextRound: (() -> Unit)? = null,
```
Comment (kingjulien1) [inline apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt]: Just like sonar indicated I would remove this unused function parameter `onNextRound`.

File: apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt
Comment (kingjulien1) [inline apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt:1]: Image

Ich hätte noch ein paar UI / Design Vorschläge für das Design des `GameResultScreen.kt`:
- Ich bin überhaupt kein fan von dem Star-Icon in der Header Sektion. Ich würde ein 🏆 (Trophy-) Emoji oder ein Trophy Icon vorschlagen. Ich wäre auch offen für irgendwelche anderen Alternativen.
- Den Hintergrund der Sektion der aufgelisteten Players (in der Abbildung als mit Rot gekennzeichnet) etwas dunkler zu gestalten damit sich der Abschnitt sich besser abhebt gegen die restlichen Sektionen.
- Die Statistiken (mit gelb umrandet) würde ich vorschlagen an die Buttons ganz unten am Screen anzu-pinnen und beim scrolling von Playern dort mit den Buttons als "sticky" zu fixieren.

File: apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt [hunk 3]
```diff
@@ -146,13 +149,52 @@ fun NavGraphBuilder.homeGraph(
...
+            // Reuse the GameViewModel from the game screen (still in backstack)
+            // so the full result data (players, scores) is available
+            val gameBackStackEntry = remember(backStackEntry) {
+                navController.previousBackStackEntry
+            }
+            val vm: GameViewModel = if (gameBackStackEntry != null) {
+                viewModel(gameBackStackEntry, factory = GenericViewModelFactory { GameViewModel(userStore) })
+            } else {
+                viewModel(factory = GenericViewModelFactory { GameViewModel(userStore) })
+            }
```
Comment (kingjulien1) [inline apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt]: I get why this is done, but I think this navigation contract is a bit fragile.

The result screen now implicitly depends on `previousBackStackEntry` still being the game screen so it can reuse the same `GameViewModel`. That works for the current flow, but it makes the result route less self-contained and more sensitive to later navigation changes. I’d at least document that assumption clearly, or consider whether the result route should eventually get the data/state it needs more explicitly.

#### A simple intermediate solution is to use a shared nav-graph-scoped `GameViewModel`.

```kotlin
navigation(
    startDestination = "${Routes.GAME}/{gameId}",
    route = Routes.GAME_FLOW
) {
    composable("${Routes.GAME}/{gameId}") { backStackEntry ->
        val parentEntry = remember(backStackEntry) {
            navController.getBackStackEntry(Routes.GAME_FLOW)
        }

        val vm: GameViewModel = viewModel(
            parentEntry,
            factory = GenericViewModelFactory { GameViewModel(userStore) }
        )

        GameScreen(vm = vm)
    }

    composable(Routes.GAME_RESULT) { backStackEntry ->
        val parentEntry = remember(backStackEntry) {
            navController.getBackStackEntry(Routes.GAME_FLOW)
        }

        val vm: GameViewModel = viewModel(
            parentEntry,
            factory = GenericViewModelFactory { GameViewModel(userStore) }
        )

        GameResultScreen(vm = vm)
    }
}
```

File: apps/android/app/src/main/java/at/aau/serg/android/core/network/RetrofitProvider.kt [hunk 4]
```diff
@@ -8,9 +8,9 @@ import java.util.concurrent.TimeUnit
 object RetrofitProvider {
 
     private val client = OkHttpClient.Builder()
-        .connectTimeout(30, TimeUnit.SECONDS)
-        .readTimeout(30, TimeUnit.SECONDS)
-        .writeTimeout(30, TimeUnit.SECONDS)
+        .connectTimeout(60, TimeUnit.SECONDS)
+        .readTimeout(60, TimeUnit.SECONDS)
+        .writeTimeout(60, TimeUnit.SECONDS)
```
Comment (kingjulien1) [inline apps/android/app/src/main/java/at/aau/serg/android/core/network/RetrofitProvider.kt]: This may be the right practical fix for the current backend hosting latency, but it is broad enough that I’d make the reason a bit more explicit.

Since this changes the global Retrofit client timeouts for everything, not just one flaky call path, I’d either add a short comment explaining why 60s is the chosen value or consider whether this should eventually be scoped more narrowly if only specific endpoints need the extra margin.

**📌 Just to be clear:** You can gladly ignore this review comment if you don't see any real significance in making the effort to completely remodel and reimplement this entire part.

File: apps/android/app/src/test/java/at/aau/serg/android/ui/screens/game/GameViewModelTest.kt
Comment (kingjulien1) [inline apps/android/app/src/test/java/at/aau/serg/android/ui/screens/game/GameViewModelTest.kt:1]: _*small remark: this comment doesnt really only link to this particular test file, but generally to all test files in this PR all the same._
 
Since this route now has both a “reuse previous game VM” path and a fallback “create a fresh VM” path, I think it would be worth locking one of those assumptions down with a focused test.

In particular, a small test around the fallback case would make it much clearer what the expected behavior is if the result screen is ever opened without the game screen still sitting in the back stack.

### Add a test for the “result route without previous game entry” fallback

If the current implementation keeps the fallback path, then that behavior should be pinned down.

**Example navigation test**
```kotlin
@Test
fun `game result route still renders when opened without previous game entry`() {
    composeTestRule.setContent {
        val navController = rememberNavController()

        NavHost(
            navController = navController,
            startDestination = Routes.GAME_RESULT
        ) {
            composable(Routes.GAME_RESULT) { backStackEntry ->
                val gameBackStackEntry = remember(backStackEntry) {
                    navController.previousBackStackEntry
                }

                val vm: GameViewModel = if (gameBackStackEntry != null) {
                    viewModel(
                        gameBackStackEntry,
                        factory = GenericViewModelFactory { GameViewModel(FakeUserStore()) }
                    )
                } else {
                    viewModel(
                        factory = GenericViewModelFactory { GameViewModel(FakeUserStore()) }
                    )
                }

                GameResultScreen(vm = vm)
            }
        }
    }

    composeTestRule
        .onNodeWithTag("game_result_screen")
        .assertExists()
}
```

For this to work nicely, the screen should expose a stable test tag:
```kotlin
Modifier.testTag("game_result_screen")
```

**Review by kingjulien1 — CHANGES_REQUESTED** [review]
Nice PR overall 👍 — the finished-game/result flow comes together well, and I only left a few smaller comments around making the route contract a bit cleaner, reducing navigation fragility, and documenting one broad network change a bit more explicitly.

If you need any help with resolving my review or in any other way or sense just tag me in a comment like so _@kingjulien1_

**Review by kingjulien1 — COMMENTED** [review]
📌 _One small note_: the test / Sonar coverage is currently a little too cozy with the 80% minimum for my personal emotional stability. If there’s an easy, low-effort way to bump it up a few points, that’d be lovely. If not, honestly just leave it — the **time and effort are probably way better wasted elsewhere**.

File: apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt (hunk 1 above)
Comment (vanidrei) [reply apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt]: Good call - the winnerId in the route was redundant since GameResultScreen only reads from uiState.gameResult anyway. Went with your suggestion and removed the route parameter entirely:

- NavigateToResult is now an object instead of a data class (no payload needed)
- Navigation call simplified to navController.navigate(Routes.GAME_RESULT)
- Route changed from "game_result/{winnerId}" to just "game_result"

File: apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt (hunk 2 above)
Comment (vanidrei) [reply apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt]: I checked again and onNextRound is actually used - it drives the "Next Round" button:
`onClick = { onNextRound?.invoke() ?: onNavigateHome()`

and in HomeGraph it's wired up with the WaitingRoom navigation when a lobbyId is available. Removing it would silently break that feature and fall back to onNavigateHome() in all cases.

What IS unused though is onShareResult — sharing is now handled internally via LocalContext + Intent.ACTION_SEND, so the callback was just dead weight. Removed that one instead.

File: apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt
Comment (vanidrei) [reply apps/android/app/src/main/java/at/aau/serg/android/ui/screens/game/GameResultScreen.kt:1]: All three points were addressed in the result screen UI rework that's part of this PR:

1. Trophy icon — replaced the star with EmojiEvents (trophy) for the winner and MilitaryTech (medal) for everyone else. The "GAME OVER" badge also uses the trophy icon now.
2. Darker player card background — introduced a dedicated ResCard color (#1D1A38) for non-highlighted player cards, noticeably darker than before.
3. Sticky stats + buttons - restructured the layout from a single LazyColumn to a Column with weight(1f) on the player list. Header, hero section, stats row, and action buttons are all fixed — only the leaderboard scrolls.

File: apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt (hunk 3 above)
Comment (vanidrei) [reply apps/android/app/src/main/java/at/aau/serg/android/ui/navigation/HomeGraph.kt]: You're right, the previousBackStackEntry assumption was fragile. Went with your suggestion - wrapped GAME and GAME_RESULT in a nested GAME_FLOW navigation graph and scoped the GameViewModel to the parent entry via navController.getBackStackEntry(Routes.GAME_FLOW). Both screens now explicitly share the same ViewModel instance through the graph scope rather than relying on back stack order.

File: apps/android/app/src/main/java/at/aau/serg/android/core/network/RetrofitProvider.kt (hunk 4 above)
Comment (vanidrei) [reply apps/android/app/src/main/java/at/aau/serg/android/core/network/RetrofitProvider.kt]: Fair point - added a comment explaining that the 60s value is tied to Render.com free-tier cold start latency and noting it should be revisited if the backend moves or if only specific endpoints end up needing it. A narrower per-client approach would be cleaner long term but overkill for now.

File: apps/android/app/src/test/java/at/aau/serg/android/ui/screens/game/GameViewModelTest.kt
Comment (vanidrei) [reply apps/android/app/src/test/java/at/aau/serg/android/ui/screens/game/GameViewModelTest.kt:1]: Good point on the missing tests. The previousBackStackEntry fallback is gone now (addressed in the nav graph refactor), so there's no fallback path to pin down anymore. 

What was actually missing were UI tests for GameResultScreen itself - added GameResultScreenTest covering: renders with null/real result, YOU WIN vs YOU FINISHED title, all players visible, home button callback, and winner name test tag. The test tag GameTestTags.RESULT_SCREEN = "game_result_screen" was already in place.

**Review by kingjulien1 — COMMENTED** [review]
@vanidrei thanks so much for this insane rework of my review - you resolved all the comments, added the tests, even upped the coverage by a lot and most of all, you added very good and detailed comment replies to my review comments  - simply amazing. 

I would actually ask you to wait with merging this PR until the accompanying / linked backend PR is finished and merged. There will be minor work to do then to switch to using the actual game metrics retrieved from the backend.

Thanks again 💯

**Review by ez-del — APPROVED** [review]
Looks good, I've played 2 rounds to test if everything works and everything seemed to work fine.

**Review by kingjulien1 — APPROVED** [review]
looks good to me - permission to merge.
